sdk/include/opentelemetry/sdk/metrics/controller.h

```diff
@@ -0,0 +1,150 @@
+#pragma once
+
+#include <atomic>
+#include <iostream>
+#include <sstream>
+#include <thread>
+#include <vector>
+#include "opentelemetry/exporters/ostream/metrics_exporter.h"
+#include "opentelemetry/metrics/instrument.h"
+#include "opentelemetry/nostd/unique_ptr.h"
+#include "opentelemetry/sdk/metrics/exporter.h"
+#include "opentelemetry/sdk/metrics/meter.h"
+#include "opentelemetry/sdk/metrics/processor.h"
+#include "opentelemetry/sdk/metrics/record.h"
+#include "opentelemetry/version.h"
+
+namespace metrics_api = opentelemetry::metrics;
+namespace trace_api   = opentelemetry::trace;
+
+OPENTELEMETRY_BEGIN_NAMESPACE
+namespace sdk
+{
+namespace metrics
+{
+
+class PushController
+{
+
+public:
+  PushController(nostd::shared_ptr<metrics_api::Meter> meter,
+                 nostd::unique_ptr<MetricsExporter> exporter,
+                 nostd::shared_ptr<MetricsProcessor> processor,
+                 double period,
+                 int timeout = 30)
+  {
+    meter_     = meter;
+    exporter_  = std::move(exporter);
+    processor_ = processor;
+    timeout_   = (unsigned int)(timeout * 1000000);  // convert seconds to microseconds
+    period_    = (unsigned int)(period * 1000000);
+  }
+
+  /*
+   * Used to check if the metrics pipeline is currecntly active
+   *
+   * @param none
+   * @return true when active, false when on standby
+   */
+  bool isActive() { return active_.load(); }
+
+  /*
+   * Begins the data processing and export pipeline.  The function first ensures that the pipeline
+   * is not already running.  If not, it begins and detaches a new thread for the Controller's run
+   * function which periodically polls the instruments for their data.
+   *
+   * @param none
+   * @return a boolean which is true when the pipeline is successfully started and false when
+   * already active
+   */
+  bool start()
+  {
+    if (!active_.load())
+    {
+      active_ = true;
+      std::thread runner(&PushController::run, this);
+      runner.detach();
+      return true;
+    }
+    return false;
+  }
+
+  /*
+   * Ends the processing and export pipeline then exports metrics one last time
+   * before returning.
+   *
+   * @param none
+   * @return none
+   */
+  void stop()
+  {
+    if (active_.load())
```

Comment:
Use `atomic_exchange` instead.

Reply:
@maxgolov Just to clarify, calls to `load()` are not thread safe?  Do  I also have to modify the way I set the values of these booleans?  (If you have time to elaborate on why the current implementation is not thread safe I would appreciate that for my own reference as well)

Reply:
The call to `load()` is thread-safe. However, the code how it is currently implemented - is not.

There is a race condition between the calls to `load()` (atomic) and the subsequent line right after, which is not atomic.

Scenario is:
- two threads both call into `start` or `stop` - either at the same time or rapidly with a very short delay.
- thread A calls `load()` - and at the same time thread B calls `load()` on a different core, e.g. both get false (system is not running yet) .... then either context switch happens right after the load returns, OR both run concurrently on multicore.
- subsequent code is non-atomically assigns the new state to `active_`, so that that sequence of two operations (load and store) as currently implemented - is NOT atomic and has a race condition.

One way of solving this as follows:

https://en.cppreference.com/w/cpp/atomic/atomic/exchange

```cpp
        virtual void start() override
        {
            if (!active_.exchange(true)) // force start, but proceed only if the previous state was 'stopped'
            {
...
            }
        };
        
        virtual void stop() override
        {
            if (active_.exchange(false)) // force stop, but proceed only if the previous state was 'started'
            {
...
            }
        };
```

Reply:
Understood, thank you very much.  I'll make a new PR for this fix.